### PR TITLE
#199 Send confirmation emails when a meeting is created

### PR DESCRIPTION
**Pull request by Ymkavi**
Screenshot 2022-09-07 114530

### CODE REVIEW

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/InformQueueService.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public class InformQueueService : IInformQueueService
         public InformQueueService(IProducerService producer, IConsumerService consumer)
         {
             _producer = producer;
-            consumer.ListenQueue();
+            consumer.Listen(null);
```
Comment (hekeemje) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/InformQueueService.cs]: why are you send to method null here? also GitHub says its bad idea 
image

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 2]
```diff
@@ -142,6 +148,45 @@ public async Task<List<OrderedMeetingTimesDto>> GetOrderedMeetingTimesAsync(long
...
+            }
+            
+            foreach (var recipient in recipients)
+            {
+                var emailDto = new EmailDto
+                {
+                    Subject = emailTemplate.Subject, 
+                    Body = emailTemplate.Body, 
+                    Recipient = recipient
+                };
```
Comment (hekeemje) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]: we can use mapper here and for Recipient  you can do .AfterMap

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Services/EmailService.cs [hunk 3]
```diff
@@ -1,44 +1,40 @@
...
 
-        public async Task<string> SendEmail(string recipient, string body, string subject)
+        public async Task<string> SendEmailAsync(string recipient, string body, string subject)
         {
-            Configuration.Default.ApiKey.Add("api-key", _apiKey);
-            var Email = new SendSmtpEmailSender(_senderName, _senderEmail);
+            var email = new SendSmtpEmailSender(_senderName, _senderEmail);
             var smtpEmailTo = new SendSmtpEmailTo(recipient);
-            List<SendSmtpEmailTo> To = new() { smtpEmailTo};
+            var to = new List<SendSmtpEmailTo> { smtpEmailTo };
```
Comment (hekeemje) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Services/EmailService.cs]: name of variable just "to" can little confuse someone who will read this code in future

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Services/EmailService.cs [hunk 4]
```diff
@@ -1,44 +1,40 @@
...
+                var sendSmtpEmail = new SendSmtpEmail(email, to, null, null, null, body, subject);
+                var result = await _api.SendTransacEmailAsync(sendSmtpEmail);
 
                 return result.MessageId;
             }
             catch (Exception e)
             {
                 _logger.LogError(e, "Email wasn't sent");
-                return string.Empty;
+                throw;
```
Comment (hekeemje) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Services/EmailService.cs:37]: why just "throw" without "throw ...." and some exception, then we will can know from exception message what went wrong

Comment (Ymkavi) [reply backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Services/EmailService.cs:37]: Here, "throw" is used to re-throw an existing exception without changing the stack trace

**Review by hekeemje — APPROVED** [review]
looks good

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/HostedServices/EmailQueueListener.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    public Task StopAsync(CancellationToken cancellationToken)
+    {
+        _consumerService.Dispose();
+        
+        return Task.CompletedTask;
+    }
+
+    private void EmailReceived(object sender, BasicDeliverEventArgs args)
+    {
+        Task.Run(async () =>
```
Comment (apofis96) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/HostedServices/EmailQueueListener.cs:38]: Use normal async here.
(Hint AsyncEventingBasicConsumer)

**Review by apofis96 — APPROVED** [review]

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/HostedServices/EmailQueueListener.cs (hunk 5 above)
Comment (apofis96) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/HostedServices/EmailQueueListener.cs]: Use normal async here.
(Hint AsyncEventingBasicConsumer)

**Review by apofis96 — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -47,6 +47,7 @@ public static void RegisterCustomServices(this IServiceCollection services, ICon
             services.AddTransient<IExternalAttendeeService, ExternalAttendeeService>();
             services.AddHttpClient<IZoomService, ZoomService>();
             services.AddRabbitMQ(configuration);
+            services.AddRabbitMqEmailSender(configuration);
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs]: we need to have 1 method for RabbitMQ registrations

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Program.cs [hunk 7]
```diff
@@ -1,19 +1,24 @@
...
 
 var builder = WebApplication.CreateBuilder(args);
 
 // Add services to the container.
 // Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
-builder.Services.AddTransient<IEmailService, EmailService>();
+builder.Services.AddRabbitMqConnection(builder.Configuration);
+builder.Services.AddEmailQueueListener(builder.Configuration);
```
Comment (tatianahutii) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Program.cs]: see the previous comment

File: backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Program.cs [hunk 8]
```diff
@@ -1,19 +1,24 @@
...
 builder.WebHost.UseUrls("http://*:5090");
 
 var app = builder.Build();
 
+Configuration.Default.ApiKey.Add("api-key", app.Configuration["SendingBlue_api_key"]);
+
```
Comment (tatianahutii) [inline backend/EasyMeets.Emailer/EasyMeets.Emailer.WebAPI/Program.cs]: why do we need this here?

File: backend/EasyMeets.Core/EasyMeets.Core.WebAPI/appsettings.json [hunk 9]
```diff
@@ -16,6 +16,12 @@
         "ExchangeType": "direct",
         "RoutingKey": "informer",
         "QueueName": "InformEveryone"
+      },
+      "EmailProducer": {
+        "ExchangeName": "Email",
+        "ExchangeType": "fanout",
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.WebAPI/appsettings.json]: Why fanout? Why do we need to send messages to all queues?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]

**Comment by Ymkavi** [comment]
"SendingBlue_api_key" environment variable has to be added

**Pull request merged**